lib/puppet/provider/route53_record.rb

```diff
@@ -0,0 +1,83 @@
+require_relative '../../puppet_x/puppetlabs/aws'
+
+class Puppet::Provider::Route53Record < PuppetX::Puppetlabs::Aws
+
+  def self.instances
+    zones_response = route53_client.list_hosted_zones()
+    records = []
+    zones_response.data.hosted_zones.each do |zone|
+      records_response = route53_client.list_resource_record_sets(hosted_zone_id: zone.id)
+      records_response.data.resource_record_sets.each do |record|
+        records << new({
+          name: record.name,
+          ensure: :present,
+          zone: zone.name,
+          ttl: record.ttl,
+          values: record.resource_records.map(&:value),
+        }) if record.type == record_type
+      end
+    end
+    records
+  end
+
+  def self.prefetch(resources)
+    instances.each do |prov|
+      if resource = resources[prov.name]
+        resource.provider = prov
+      end
+    end
+  end
+
+  def exists?
+    Puppet.info("Checking if #{self.class.record_type} record #{name} exists")
+    @property_hash[:ensure] == :present
+  end
+
+  def record_hash(action)
+    records = resource[:values] ? resource[:values].map { |v| {value: v} } : []
+    {
+      hosted_zone_id: zone_id,
+      change_batch: {
+        changes: [{
+          action: action,
+          resource_record_set: {
+            name: resource[:name],
+            type: self.class.record_type,
+            ttl: resource[:ttl],
+            resource_records: records,
+          }
+        }]
+      }
+    }
+  end
+
+  def zone_id
+    zones = route53_client.list_hosted_zones.data.hosted_zones.select { |zone| zone.name == resource[:zone] }
```

Comment:
This line is a bit long, splitting it into a do/end block would be a little nicer.